include jvm-options change in config change integration test

jsanda opened the pull request:
Note that this is a WIP. 

I thought it would be good to have test coverage for changes to `jvm-options`. It also serves as documentation. I had to spend time hunting through the [cass-config-definitions](https://github.com/datastax/cass-config-definitions) and the [cass-config-builder](https://github.com/datastax/cass-config-builder) repos to figure out how to make some changes.

I have some local changes not included with this commit that were necessary to get the test working with Cassandra 3.11.x (see #25 and #26 for details). The test in master currently only works with DSE.

For this PR I would like to update the existing to to include the `jvm-options` change and then add a variation to test against Cassandra. Given how the integration tests are structured I am not sure how best to proceed.

jsanda commented:
Is the CI set up to accept PRs from forks?

Here is there error I found - `Error: MO_ECR_REPO is a required environment variable`

And the github [workflow](https://github.com/datastax/cass-operator/blob/master/.github/workflows/operatorPRBuildAndDeploy.yml#L28) has this:

```yaml
      - name: Deploy to ECR
        if: github.repository == 'datastax/cass-operator'
        env:
          MO_ECR_ID: ${{ secrets.ECR_ID }}
          MO_ECR_SECRET: ${{ secrets.ECR_SECRET }}
          MO_ECR_REPO: ${{ secrets.ECR_REPO }}
```

jimdickinson commented on tests/config_change/config_change_suite_test.go:
```diff
@@ -94,7 +94,7 @@ var _ = Describe(testName, func() {
 
 			step = "change the config"
-			json = "{\"spec\": {\"config\": {\"cassandra-yaml\": {\"file_cache_size_in_mb\": 123}}}}"
+			json = "{\"spec\": {\"config\": {\"cassandra-yaml\": {\"file_cache_size_in_mb\": 123}, \"jvm-options\": {\"garbage_collector\": \"CMS\"}}}}"
```
should be jvm-server-options for DSE though?

jsanda replied:
Yes, it should be `jvm-server-options`. I am in a bit of a catch 22 situation. I cannot pull the DSE images (see #25). I can only run the test against Cassandra 3.11.x. I will change that back to `jvm-server-options` prior to if/when the PR is otherwise ready to merge.

sandoichi commented:
> Is the CI set up to accept PRs from forks?
> 
> Here is there error I found - `Error: MO_ECR_REPO is a required environment variable`
> 
> And the github [workflow](https://github.com/datastax/cass-operator/blob/master/.github/workflows/operatorPRBuildAndDeploy.yml#L28) has this:
> 
> ```yaml
>       - name: Deploy to ECR
>         if: github.repository == 'datastax/cass-operator'
>         env:
>           MO_ECR_ID: ${{ secrets.ECR_ID }}
>           MO_ECR_SECRET: ${{ secrets.ECR_SECRET }}
>           MO_ECR_REPO: ${{ secrets.ECR_REPO }}
> ```

Nope, forks cannot access the secrets of the repo.

> Is the CI set up to accept PRs from forks?
> 
> Here is there error I found - `Error: MO_ECR_REPO is a required environment variable`
> 
> And the github [workflow](https://github.com/datastax/cass-operator/blob/master/.github/workflows/operatorPRBuildAndDeploy.yml#L28) has this:
> 
> ```yaml
>       - name: Deploy to ECR
>         if: github.repository == 'datastax/cass-operator'
>         env:
>           MO_ECR_ID: ${{ secrets.ECR_ID }}
>           MO_ECR_SECRET: ${{ secrets.ECR_SECRET }}
>           MO_ECR_REPO: ${{ secrets.ECR_REPO }}
> ```

Nope, forks cannot access repo secrets. It looks like you have an outdated github action workflow in your branch. If you merge the latest from master, it will have a guard against attempting to deploy to ECR from your fork.

jsanda commented:
>Nope, forks cannot access repo secrets. It looks like you have an outdated github action workflow in your branch. If you merge the latest from master, it will have a guard against attempting to deploy to ECR from your fork.

What should I do then about to this PR, and any other future PRs?

sandoichi commented:
> > Nope, forks cannot access repo secrets. It looks like you have an outdated github action workflow in your branch. If you merge the latest from master, it will have a guard against attempting to deploy to ECR from your fork.
> 
> What should I do then about to this PR, and any other future PRs?

If you merge the latest master branch into your fork, you will get the updated workflow that won't error out. It will still run unit tests, build the code, and build the docker image. It just won't attempt to publish to ECR.

jsanda commented:
>If you merge the latest master branch into your fork, you will get the updated workflow that won't error out. It will still run unit tests, build the code, and build the docker image. It just won't attempt to publish to ECR.

Awesome, thanks!

sandoichi commented:
> > If you merge the latest master branch into your fork, you will get the updated workflow that won't error out. It will still run unit tests, build the code, and build the docker image. It just won't attempt to publish to ECR.
> 
> Awesome, thanks!

Hey, it looks like it still tried to run the ECR push for some reason even after you got the latest code with that guard in place... I'm going to investigate and see what's going on with that. Sorry for the confusion.

jsanda commented:
Since #30 has been fixed, I went ahead and rebased and did a force push on my branch.

I also reverted the property name back to `jvm-server-options` for DSE. 

Since there are differences between DSE and Cassandra and how they are handled, I would like to add a similar integration that uses OSS C* instead of DSE. That brings me back to a couple of questions:

1. Would it be reasonable to add a test that targets OSS Cassandra?
2. If the answer to 1. is yes, then where should that test live?

sandoichi commented:
> Since #30 has been fixed, I went ahead and rebased and did a force push on my branch.
> 
> I also reverted the property name back to `jvm-server-options` for DSE.
> 
> Since there are differences between DSE and Cassandra and how they are handled, I would like to add a similar integration that uses OSS C* instead of DSE. That brings me back to a couple of questions:
> 
>     1. Would it be reasonable to add a test that targets OSS Cassandra?
> 
>     2. If the answer to 1. is yes, then where should that test live?

We currently have 1 integration test that targets OSS [here](https://github.com/datastax/cass-operator/blob/master/tests/oss_test_all_the_things/oss_test_all_the_things_suite_test.go). 

If there is a need to add another one, it can be done alongside the existing ones. We don't differentiate between the DSE/OSS from a test discovery/test running standpoint. The test yaml file for the OSS test also lives alongside the regular DSE ones [here](https://github.com/datastax/cass-operator/blob/master/tests/testdata/oss-three-rack-three-node-dc.yaml)
